course-material/intro-to-sql/connecting-to-databases.md

```diff
@@ -1,14 +1,606 @@
+
+We next group by their reason for absence. 
+
+`HAVING COUNT() > (`  begins the condition that the groups must meet to be included in the results. Only groups where the count of rows (representing the number of instances of each `Reason_for_absence` among social drinkers) is greater than a certain value will be included.
+
+`SELECT COUNT() FROM absenteeism WHERE Social_drinker = 0 GROUP BY Reason_for_absence)`  is a subquery that calculates the count of rows for each `Reason_for_absence` where `Social_drinker` is 0 (indicating the employee is not a social drinker), effectively giving us the number of instances of each `Reason_for_absence` among non-social drinkers.
+
+```python
+%%sql
+SELECT Reason_for_absence 
+FROM absenteeism 
+WHERE Social_drinker = 1 
+GROUP BY Reason_for_absence 
+HAVING COUNT() > (
+    SELECT COUNT() 
+    FROM absenteeism 
+    WHERE Social_drinker = 0 
+    GROUP BY Reason_for_absence);
+
+```
+</details>
+<!-- #endregion -->
+
+## Bonus
+
+### In-memory Database with DuckDB
+
+Although URL-based connections are more secure, can handle various types of workloads, and offer more functionality, in-memory databases are a great option for quick querying and testing. In this tutorial, we'll use [DuckDB](https://jupysql.ploomber.io/en/latest/integrations/duckdb.html) to create an in-memory database with JupySQL.
+
+The first step is to install the dependencies:
+
+```{code-cell} ipython3
+%pip install jupysql duckdb duckdb-engine --quiet
+```
+
+Then, load the ipython-sql library using the `%load_ext` iPython extension syntax and connect to the database:
+
+```{code-cell} ipython3
+:tags: [remove-output]
+
+%load_ext sql
+```
+
+Finally, load `%sql` and initialize the database:
+
+```{code-cell} ipython3
+%sql duckdb://
+```
+
+Download some data:
+
+```{code-cell} ipython3
+import urllib
+
+urllib.request.urlretrieve(
+    "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv",
+    "penguins.csv",
+)
+```
+
+You're all set!
+
+```{code-cell} ipython3
+%sql select * from penguins.csv limit 3
+```
+
+### Databases Supported by JupySQL
+
+Check out our guides for connecting to supported databases:
+
+- [PostgreSQL](https://jupysql.ploomber.io/en/latest/integrations/postgres-connect.html)
+- [ClickHouse](https://jupysql.ploomber.io/en/latest/integrations/clickhouse.html)
+- [MariaDB](https://jupysql.ploomber.io/en/latest/integrations/mariadb.html)
+- [MindsDB](https://jupysql.ploomber.io/en/latest/integrations/mindsdb.html)
+- [MSSQL](https://jupysql.ploomber.io/en/latest/integrations/mssql.html)
+- [MySQL](https://jupysql.ploomber.io/en/latest/integrations/mysql.html)
+- [QuestDB](https://jupysql.ploomber.io/en/latest/integrations/questdb.html)
+- [Oracle](https://jupysql.ploomber.io/en/latest/integrations/oracle.html)
+- [Trino](https://jupysql.ploomber.io/en/latest/integrations/trinodb.html)
+
++++
```

Comment:
Suggestion: add one to three hands on questions that allow people to practice their skills. 

For example:

1. Form a connection string for a postgres database
2. Form a connection string for a DuckDB instance
3. What is the difference between using MySQL and PostgreSQL
4. Load a csv into a DuckDB instance

The questions above are suggestions. The goal is to help them practice via the Google colab. You can use a similar format as the absenteeism dataset where you have an open ended question, you leave a cell open for their answer, and using markdown you can hide the answer.

Reply:
Understood, thanks for the detailed feedback, incorporating them now and committing soon.